mslib/msui/wms_control.py

```diff
@@ -547,51 +554,83 @@ def initialise_wms(self, base_url, version="1.3.0"):
         username, password = constants.WMS_LOGIN_CACHE.get(base_url, (None, None))
 
-        try:
-            str(base_url)  # to provoke early Unicode Error
-            while wms is None:
-                try:
-                    wms = MSSWebMapService(base_url, version=version,
-                                           username=username, password=password)
-                except owslib.util.ServiceException as ex:
-                    logging.error("ERROR: %s %s", type(ex), ex)
-                    if str(ex).startswith("401") or str(ex).find("Error 401") >= 0 or str(ex).find(
-                            "Unauthorized") >= 0:
-                        # Catch the "401 Unauthorized" error if one has been
-                        # returned by the server and ask the user for username
-                        # and password.
-                        # WORKAROUND (mr, 28Mar2013) -- owslib doesn't recognize
-                        # the Apache 401 messages, we get an XML message here but
-                        # no error code. Quick workaround: Scan the XML message for
-                        # the string "Error 401"...
-                        dlg = MSS_WMS_AuthenticationDialog(parent=self.multilayers)
-                        dlg.setModal(True)
-                        if dlg.exec_() == QtWidgets.QDialog.Accepted:
-                            username, password = dlg.getAuthInfo()
-                            # If user & pw have been entered, cache them.
-                            constants.WMS_LOGIN_CACHE[base_url] = (username, password)
-                        else:
-                            break
+        def on_success(wms):
+            self.cpdlg.setValue(9)
+            if wms is not None:
+
+                # update the combo box, if entry requires change/insertion
+                found = False
+                current_url = self.multilayers.cbWMS_URL.currentText()
+                for count in range(self.multilayers.cbWMS_URL.count()):
+                    if self.multilayers.cbWMS_URL.itemText(count) == current_url:
+                        self.multilayers.cbWMS_URL.setItemText(count, base_url)
+                        self.multilayers.cbWMS_URL.setCurrentIndex(count)
+                        found = True
+                        break
+                    if self.multilayers.cbWMS_URL.itemText(count) == current_url:
+                        self.multilayers.cbWMS_URL.setCurrentIndex(count)
+                        found = True
+                if not found:
+                    logging.debug("inserting URL: %s", base_url)
+                    add_wms_urls(self.multilayers.cbWMS_URL, [base_url])
+                    self.multilayers.cbWMS_URL.setEditText(base_url)
+                    save_settings_qsettings('wms', {'recent_wms_url': base_url})
+
+                self.activate_wms(wms)
+                WMS_SERVICE_CACHE[wms.url] = wms
+                self.cpdlg.close()
+
+        def on_failure(e):
```

Comment:
I'm not sure  if it gets better readable later for me if instead of (e) we have here (wms) as used for the on_sucess() too.

Am I correct that it is derivated from Exception?

Reply:
The failed signal returns an Exception. I thought e would further signify this.

Reply:
yes it does

I compared the two functions (top down reading) 

I should review my comments than bottom up.